spanner/mutation.go

```diff
@@ -470,41 +469,39 @@ func (m Mutation) proto() (*sppb.Mutation, error) {
 // mutationsProto turns a spanner.Mutation array into a sppb.Mutation array,
 // it is convenient for sending batch mutations to Cloud Spanner.
 func mutationsProto(ms []*Mutation) ([]*sppb.Mutation, *sppb.Mutation, error) {
-	var selectedMutation *Mutation
-	var nonInsertMutations []*Mutation
-
-	l := make([]*sppb.Mutation, 0, len(ms))
-	for _, m := range ms {
-		if m.op != opInsert {
-			nonInsertMutations = append(nonInsertMutations, m)
-		}
-		if selectedMutation == nil {
-			selectedMutation = m
-		}
-		// Track the INSERT mutation with the highest number of values if only INSERT mutation were found
-		if selectedMutation.op == opInsert && m.op == opInsert && len(m.values) > len(selectedMutation.values) {
-			selectedMutation = m
-		}
-
-		// Convert the mutation to sppb.Mutation and add to the list
+	n := len(ms)
+	out := make([]*sppb.Mutation, 0, n)
+	if n == 0 {
+		return out, nil, nil
+	}
+	maxInsertIdx := -1
+	maxInsertVals := -1
+	nonInsertCount := 0
+	selectedNonInsertIdx := -1
+	for i, m := range ms {
 		pb, err := m.proto()
 		if err != nil {
 			return nil, nil, err
 		}
-		l = append(l, pb)
+		out = append(out, pb)
+		if m.op == opInsert {
+			if v := len(m.values); v >= maxInsertVals {
+				maxInsertVals, maxInsertIdx = v, i
+			}
+			continue
+		}
+		nonInsertCount++
+		if rand.Intn(nonInsertCount) == 0 {
+			selectedNonInsertIdx = i
+		}
```

Comment:
nit: at first glance it's a bit surprising that this ends up being a uniform distribution, so perhaps a comment can help here:
```suggestion
         // Pick a non-insert mutation uniformly at random.
		if rand.Intn(nonInsertCount) == 0 {
			selectedNonInsertIdx = i
		}
```